src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,682 @@
+               const ignition::msgs::PointCloudPacked &_msg);
+
+  /// \brief Unpacks the point cloud retrieved by the Depth Camera
+  /// into XYZ data
+  /// \param[in] _msg Message containing the data to unpack
+  public: void UnpackPointCloudMsg(
+               const ignition::msgs::PointCloudPacked &_msg);
+
+  /// \brief Computes the normal forces of the Optical Tactile sensor
+  /// \param[in] _msg Message returned by the Depth Camera
+  /// \param[in] _visualizeForces Whether to visualize the forces or not
+  ///
+  /// Implementation inspired by
+  /// https://stackoverflow.com/questions/
+  /// 34644101/calculate-surface-normals-from-depth-image-
+  /// using-neighboring-pixels-cross-produc
+  public: void ComputeNormalForces(
+               const ignition::msgs::PointCloudPacked &_msg,
+               const bool _visualizeForces);
+
+  /// \brief Resolution of the sensor in pixels to skip.
+  public: int resolution{100};
+
+  /// \brief Whether to visualize the normal forces.
+  public: int visualize_forces{false};
+
+  /// \brief Model interface.
+  public: Model model{kNullEntity};
+
+  /// \brief Transport node to visualize data on Gazebo.
+  public: transport::Node node;
+
+  /// \brief Collision entities that have been designated as contact sensors.
+  public: std::vector<Entity> sensorEntities;
+
+  /// \brief Collision entities of the simulation.
+  public: std::vector<Entity> collisionEntities;
+
+  /// \brief Filtered collisions from the simulation that belong to
+  /// one or more sensors.
+  public: std::vector<Entity> filteredCollisionEntities;
+
+  /// \brief Wheter the plugin is enabled.
+  public: bool enabled{true};
+
+  /// \brief Initialization flag.
+  public: bool initialized{false};
+
+  /// \brief Copy of the sdf configuration used for this plugin.
+  public: sdf::ElementPtr sdfConfig;
+
+  /// \brief Name of the model to which the sensor is attached.
+  public: std::string modelName;
+
+  /// \brief Message for visualizing contact positions
+  public: ignition::msgs::Marker positionMarkerMsg;
+
+  /// \brief Radius of the visualized contact sphere
+  public: double contactRadius{0.10};
+
+  /// \brief Message for visualizing contact forces
+  public: ignition::msgs::Marker forceMarkerMsg;
+
+  /// \brief Length of the visualized force cylinder
+  public: double forceLenght{0.20};
+
+  /// \brief Position interpolated from the Contact messages
+  public: std::vector<ignition::math::Vector3d> interpolatedPosition;
+
+  /// \brief Pose of the sensor model
+  public: ignition::math::Pose3f sensorWorldPose;
+
+  /// \brief Whether a new message has been returned by the Depth Camera
+  public: bool newCameraMsg{false};
+
+  /// \brief Depth Camera update rate
+  public: float cameraUpdateRate{1};
+
+  /// \brief Message returned by the Depth Camera
+  public: ignition::msgs::PointCloudPacked cameraMsg;
+
+  /// \brief 3D vector to store unpacked XYZ points
+  std::vector<std::vector<std::vector<float>>> imageXYZ;
```

Comment:
I had tried `ignition::common::Image` but [it is not](https://github.com/ignitionrobotics/ign-common/blob/master/graphics/src/Image.cc#L191) compatible with the data format of `ignition::msgs::PointCloudPacked`. 
What I can do is work with a raw pointer and somehow map the (i, j, XYZ) to the pointer position (basically what I'm doing [here](https://github.com/mcres/ign-gazebo/blob/optical_tactile_plugin/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L565)). That way I would get rid of `UnpackPointCloudMsg()` and I'd just run `ComputeNormalForces()`.

Reply:
That sounds about right. The file has changed quite a bit since the comment and I'm not seeing the line you meant. Could you point to the line again?

@azeey Is there something equivalent to Eigen::MatrixXf that can be used to manipulate a matrix of size width * height * 3, where width and height are passed in by the user? Or does it need to be a raw array with manual indexing?

Reply:
Yup, here's the updated link: https://github.com/mcres/ign-gazebo/blob/optical_tactile_plugin/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L807. Instead of the `imageXYZ[i][j][k]` vector when computing `dxdi` and `dxdj`, there would be a mapping function instead, for instance `imageXYZ(i,j,k)`.

Reply:
You should be able to use Eigen since ign-gazebo depends on it indirectly via the eigen3 component of ign-math6. We already use it in the multicopter_control system. If you do use Eigen, be sure to specify the appropriate [storage option](https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html) (RowMajor or ColumnMajor) to improve performance

Reply:
It seems that we have two options here:

1. Using MatrixXf: I would have to unpack the data anyways but it would be more efficient than a nested vector (I assume?).
2. Manually indexing the raw array returned in the camera msg: I wouldn't have to unpack the data, but I would have to call the previously mentioned indexing function each time I want to access the value.

I'd go with no. 2, what do you think?

Reply:
@mabelzhang, I think I was too quick to answer about `MatrixXf` because I now realize that you really want a ndarray/tensor, which exists in Eigen, but is unsupported. @mcres, option 2 makes sense. I don't think you can get more efficient than calculating the index into the raw array in the point cloud message. Copying them into a new 3d std::vector incurs allocation costs and the allocated memory is unlikely to be contiguous between dimensions, which would result in unnecessary cache misses.